From 4a0fa4b45d9c35adb80be0c2bb80b471561fdd15 Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Sun, 12 Dec 2021 20:48:26 -0300
Subject: [PATCH] Disable InventoryMoveItemEvent


diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index 3d15433b6..d11fe2393 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -267,6 +267,10 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
 
                         // CraftBukkit start - Call event when pushing items into other inventories
                         CraftItemStack oitemstack = CraftItemStack.asCraftMirror(this.splitStack(i, world.spigotConfig.hopperAmount)); // Spigot
+                        // FlamePaper start - add an option to turn of InventoryMoveItemEvent
+                        final org.bukkit.inventory.ItemStack stack;
+                        if (HopperHelper.isFireInventoryMoveItemEvent(this)) {
+                        // FlamePaper end
 
                         Inventory destinationInventory;
                         // Have to special case large chests as they work oddly
@@ -283,11 +287,18 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                             this.d(world.spigotConfig.hopperTransfer); // Spigot
                             return false;
                         }
-                        int origCount = event.getItem().getAmount(); // Spigot
-                        ItemStack itemstack1 = addItem(iinventory, CraftItemStack.asNMSCopy(event.getItem()), enumdirection);
+                        // TacoSpigot start
+                        stack = event.getItem();
+                        // handle cases where the event is not fired
+                        } else {
+                            stack = oitemstack;
+                        }
+                        int origCount = stack.getAmount(); // Spigot
+                        ItemStack itemstack1 = addItem(iinventory, CraftItemStack.asNMSCopy(stack), enumdirection);
+                        // TacoSpigot end
 
                         if (itemstack1 == null || itemstack1.count == 0) {
-                            if (event.getItem().equals(oitemstack)) {
+                            if (stack.equals(oitemstack)) { // FlamePaper - event.getItem() -> stack
                                 iinventory.update();
                             } else {
                                 this.setItem(i, itemstack);
@@ -420,6 +431,11 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
             // CraftBukkit start - Call event on collection of items from inventories into the hopper
             CraftItemStack oitemstack = CraftItemStack.asCraftMirror(iinventory.splitStack(i, ihopper.getWorld().spigotConfig.hopperAmount)); // Spigot
 
+            // FlamePaper start - add an option to turn of InventoryMoveItemEvent
+            final org.bukkit.inventory.ItemStack stack;
+            if (HopperHelper.isFireInventoryMoveItemEvent(ihopper)) {
+            // FlamePaper end
+
             Inventory sourceInventory;
             // Have to special case large chests as they work oddly
             if (iinventory instanceof InventoryLargeChest) {
@@ -441,11 +457,18 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                 }
                 return false;
             }
-            int origCount = event.getItem().getAmount(); // Spigot
-            ItemStack itemstack2 = addItem(ihopper, CraftItemStack.asNMSCopy(event.getItem()), null);
+            // FlamePaper start
+            stack = event.getItem();
+            // handle cases where the event is not fired
+            } else {
+                stack = oitemstack;
+            }
+            int origCount = stack.getAmount(); // Spigot
+            ItemStack itemstack2 = addItem(ihopper, CraftItemStack.asNMSCopy(stack), null);
+            // FlamePaper end
 
             if (itemstack2 == null || itemstack2.count == 0) {
-                if (event.getItem().equals(oitemstack)) {
+                if (stack.equals(oitemstack)) { // FlamePaper
                     iinventory.update();
                 } else {
                     iinventory.setItem(i, itemstack1);
diff --git a/src/main/java/org/github/paperspigot/HopperHelper.java b/src/main/java/org/github/paperspigot/HopperHelper.java
index e43cb912c..2ced846b5 100644
--- a/src/main/java/org/github/paperspigot/HopperHelper.java
+++ b/src/main/java/org/github/paperspigot/HopperHelper.java
@@ -1,15 +1,21 @@
 package org.github.paperspigot;
 
+import org.bukkit.event.inventory.InventoryMoveItemEvent;
+
 import net.minecraft.server.Block;
 import net.minecraft.server.BlockChest;
 import net.minecraft.server.BlockPosition;
 import net.minecraft.server.Blocks;
+import net.minecraft.server.IHopper;
 import net.minecraft.server.IInventory;
 import net.minecraft.server.TileEntity;
 import net.minecraft.server.TileEntityHopper;
 import net.minecraft.server.World;
 
 public class HopperHelper {
+    public static boolean isFireInventoryMoveItemEvent(IHopper hopper) {
+        return PaperSpigotConfig.isHopperFireIMIE && InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length > 0;
+    }
 
     public static TileEntityHopper getHopper(World world, BlockPosition pos) {
         if (world.getType(pos).getBlock() != Blocks.HOPPER) return null;
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
index 85dda6fdc..b23357e01 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
@@ -186,6 +186,11 @@ public class PaperSpigotConfig
         bookMaxPages = getInt("book.max_pages", 5);
     }
     
+    public static boolean isHopperFireIMIE;
+    private static void isHopperFireIMIE() {
+        isHopperFireIMIE = getBoolean("hopper.fire-inventory-move-item-event", true);
+    }
+
     public static boolean isHopperPushBased;
     private static void isHopperPushBased()
     {
-- 
2.32.0

