From c9e6d8b1a87f6932b7015ab99d9e2211a28bcadd Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Sun, 12 Dec 2021 19:09:13 -0300
Subject: [PATCH] Push based hoppers


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 1f8b1c726..3227e0d31 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -3,9 +3,16 @@ package net.minecraft.server;
 import java.util.Iterator;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.github.paperspigot.HopperPusher; // FlamePaper
 import org.bukkit.event.player.PlayerPickupItemEvent; // CraftBukkit
 
-public class EntityItem extends Entity {
+// FlamePaper start - implement HopperPusher
+public class EntityItem extends Entity implements HopperPusher {
+    @Override
+    public boolean acceptItem(TileEntityHopper hopper) {
+        return TileEntityHopper.a(hopper, this);
+    }
+    // FlamePaper end
 
     private static final Logger b = LogManager.getLogger();
     private int age;
@@ -59,6 +66,7 @@ public class EntityItem extends Entity {
             this.die();
         } else {
             super.t_();
+            if (tryPutInHopper()) return; // TacoSpigot
             // CraftBukkit start - Use wall time for pickup and despawn timers
             int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
             if (this.pickupDelay != 32767) this.pickupDelay -= elapsedTicks;
@@ -124,6 +132,7 @@ public class EntityItem extends Entity {
     // Spigot start - copied from above
     @Override
     public void inactiveTick() {
+        if (tryPutInHopper()) return; // FlamePaper
         // CraftBukkit start - Use wall time for pickup and despawn timers
         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
         if (this.pickupDelay != 32767) this.pickupDelay -= elapsedTicks;
diff --git a/src/main/java/net/minecraft/server/EntityMinecartContainer.java b/src/main/java/net/minecraft/server/EntityMinecartContainer.java
index c6487f216..54dc6a9ba 100644
--- a/src/main/java/net/minecraft/server/EntityMinecartContainer.java
+++ b/src/main/java/net/minecraft/server/EntityMinecartContainer.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.github.paperspigot.HopperPusher; // FlamePaper
+
 // CraftBukkit start
 import java.util.List;
 
@@ -8,7 +10,14 @@ import org.bukkit.entity.HumanEntity;
 import org.bukkit.inventory.InventoryHolder;
 // CraftBukkit end
 
-public abstract class EntityMinecartContainer extends EntityMinecartAbstract implements ITileInventory {
+// FlamePaper start - HopperPusher
+public abstract class EntityMinecartContainer extends EntityMinecartAbstract implements ITileInventory, HopperPusher {
+    
+    @Override
+    public boolean acceptItem(TileEntityHopper hopper) {
+        return TileEntityHopper.acceptItem(hopper, this);
+    }
+    // FlamePaper end
 
     private ItemStack[] items = new ItemStack[27]; // CraftBukkit - 36 -> 27
     private boolean b = true;
diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index 760a0767e..3d15433b6 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -3,6 +3,9 @@ package net.minecraft.server;
 import java.util.Iterator;
 import java.util.List;
 
+import org.github.paperspigot.HopperHelper; // FlamePaper
+import org.github.paperspigot.PaperSpigotConfig; // FlamePaper
+
 // CraftBukkit start
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
@@ -200,7 +203,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                 }
             }
             // PaperSpigot start
-            if (world.paperSpigotConfig.useHopperCheck && !this.n()) {
+            if (world.paperSpigotConfig.useHopperCheck && !PaperSpigotConfig.isHopperPushBased && !this.n()) { // FlamePaper - dont use hopper check in push mode
                 this.d(world.spigotConfig.hopperCheck);
             }
             // PaperSpigot end
@@ -225,6 +228,12 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
         return true;
     }
 
+    // FlamePaper start
+    public boolean canAcceptItems() {
+        return !this.n() && !this.q() && BlockHopper.f(this.u());
+    }
+    // FlamePaper end
+
     private boolean q() {
         ItemStack[] aitemstack = this.items;
         int i = aitemstack.length;
@@ -241,7 +250,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
     }
 
     private boolean r() {
-        IInventory iinventory = this.H();
+        IInventory iinventory = HopperHelper.getInventory(getWorld(), getPosition().shift(BlockHopper.b(this.u()))); // FlamePaper
 
         if (iinventory == null) {
             return false;
@@ -346,8 +355,21 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
         return true;
     }
 
+    // PaperSpigot start - Split methods, one that pushes and one that pulls
+    @Deprecated
     public static boolean a(IHopper ihopper) {
-        IInventory iinventory = b(ihopper);
+        IInventory iinventory;
+        if (PaperSpigotConfig.isHopperPushBased && ihopper instanceof TileEntityHopper) {
+            BlockPosition pos = ((TileEntityHopper) ihopper).getPosition().up(); // Only pull from a above, because everything else comes to us
+            iinventory = HopperHelper.getInventory(ihopper.getWorld(), pos);
+        } else {
+            iinventory = b(ihopper); // Use old behavior for BB entity searching
+        }
+        return acceptItem(ihopper, iinventory);
+    }
+ 
+    public static boolean acceptItem(IHopper ihopper, IInventory iinventory) {
+    // FlamePaper end
 
         if (iinventory != null) {
             EnumDirection enumdirection = EnumDirection.DOWN;
@@ -374,7 +396,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                     }
                 }
             }
-        } else {
+        } else if (!PaperSpigotConfig.isHopperPushBased || !(ihopper instanceof TileEntityHopper)) { // FlamePaper - only search for entities in 'pull mode'
             Iterator iterator = a(ihopper.getWorld(), ihopper.A(), ihopper.B() + 1.0D, ihopper.C()).iterator();
 
             while (iterator.hasNext()) {
@@ -579,6 +601,16 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
         return (IInventory) object;
     }
 
+    // FlamePaper start
+    public AxisAlignedBB getHopperLookupBoundingBox() {
+        return getHopperLookupBoundingBox(this.A(), this.B() + 1.0D, this.C());
+    }
+
+    private static AxisAlignedBB getHopperLookupBoundingBox(double d0, double d1, double d2) {
+        return new AxisAlignedBB(d0 - 0.5D, d1 - 0.5D, d2 - 0.5D, d0 + 0.5D, d1 + 0.5D, d2 + 0.5D);
+    }
+    // FlamePaper end
+
     private static boolean a(ItemStack itemstack, ItemStack itemstack1) {
         return itemstack.getItem() != itemstack1.getItem() ? false : (itemstack.getData() != itemstack1.getData() ? false : (itemstack.count > itemstack.getMaxStackSize() ? false : ItemStack.equals(itemstack, itemstack1)));
     }
diff --git a/src/main/java/org/github/paperspigot/HopperHelper.java b/src/main/java/org/github/paperspigot/HopperHelper.java
new file mode 100644
index 000000000..e43cb912c
--- /dev/null
+++ b/src/main/java/org/github/paperspigot/HopperHelper.java
@@ -0,0 +1,35 @@
+package org.github.paperspigot;
+
+import net.minecraft.server.Block;
+import net.minecraft.server.BlockChest;
+import net.minecraft.server.BlockPosition;
+import net.minecraft.server.Blocks;
+import net.minecraft.server.IInventory;
+import net.minecraft.server.TileEntity;
+import net.minecraft.server.TileEntityHopper;
+import net.minecraft.server.World;
+
+public class HopperHelper {
+
+    public static TileEntityHopper getHopper(World world, BlockPosition pos) {
+        if (world.getType(pos).getBlock() != Blocks.HOPPER) return null;
+        TileEntity tileEntity = world.getTileEntity(pos);
+        if (tileEntity instanceof TileEntityHopper) {
+            return (TileEntityHopper) tileEntity;
+        }
+        return null;
+    }
+
+    public static IInventory getInventory(World world, BlockPosition position) {
+        Block block = world.getType(position).getBlock();
+        if (block instanceof BlockChest) {
+            return ((BlockChest) block).f(world, position);
+        }
+        if (block.isTileEntity()) {
+            TileEntity tile = world.getTileEntity(position);
+            if (tile instanceof IInventory) return (IInventory) tile;
+        }
+        return null;
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/java/org/github/paperspigot/HopperPusher.java b/src/main/java/org/github/paperspigot/HopperPusher.java
new file mode 100644
index 000000000..aff91a86b
--- /dev/null
+++ b/src/main/java/org/github/paperspigot/HopperPusher.java
@@ -0,0 +1,65 @@
+package org.github.paperspigot;
+
+import net.minecraft.server.AxisAlignedBB;
+import net.minecraft.server.BlockPosition;
+import net.minecraft.server.Entity;
+import net.minecraft.server.IHopper;
+import net.minecraft.server.TileEntityHopper;
+import net.minecraft.server.World;
+
+public interface HopperPusher {
+
+    public default TileEntityHopper findHopper() {
+        BlockPosition pos = new BlockPosition(getX(), getY(), getZ());
+        int startX = pos.getX() - 1;
+        int endX = pos.getX() + 1;
+        int startY = Math.max(0, pos.getY() - 1);
+        int endY = Math.min(255, pos.getY() + 1);
+        int startZ = pos.getZ() - 1;
+        int endZ = pos.getZ() + 1;
+        for (int x = startX; x <= endX; x++) {
+            for (int y = startY; y <= endY; y++) {
+                for (int z = startZ; z <= endZ; z++) {
+                    BlockPosition adjacentPos = new BlockPosition(x, y, z);
+                    TileEntityHopper hopper = HopperHelper.getHopper(getWorld(), adjacentPos);
+                    if (hopper == null) continue; // Avoid playing with the bounding boxes, if at all possible
+                    /**
+                     * We add one to getY(), just like {@link TileEntityHopper#b(IHopper)}
+                     */
+                    AxisAlignedBB hopperBoundingBox = hopper.getHopperLookupBoundingBox();
+                    if (hopperBoundingBox.b(this.getBoundingBox())) { // AxisAlignedBB.b(AxisAlignedBB) -> isIntersect()
+                        return hopper;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    public boolean acceptItem(TileEntityHopper hopper);
+
+    public default boolean tryPutInHopper() {
+        if (!PaperSpigotConfig.isHopperPushBased) return false;
+        TileEntityHopper hopper = findHopper();
+        return hopper != null && hopper.canAcceptItems() && acceptItem(hopper);
+    }
+
+    public AxisAlignedBB getBoundingBox();
+
+    public World getWorld();
+
+    // Default implementations for entities
+
+    public default double getX() {
+        return ((Entity) this).locX;
+    }
+
+    public default double getY() {
+        return ((Entity) this).locY;
+    }
+
+    public default double getZ() {
+        return ((Entity) this).locZ;
+    }
+
+}
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
index 488769244..85dda6fdc 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
@@ -186,6 +186,12 @@ public class PaperSpigotConfig
         bookMaxPages = getInt("book.max_pages", 5);
     }
     
+    public static boolean isHopperPushBased;
+    private static void isHopperPushBased()
+    {
+        isHopperPushBased = getBoolean( "hopper.push-based", true );
+    }
+
     public static double knockbackFriction;
     public static double knockbackHorizontal;
     public static double knockbackVertical;
-- 
2.32.0

