From 13d1b104b20c2d3285e3a35ba98425d7c4b1ab10 Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Sun, 13 Mar 2022 22:22:09 -0300
Subject: [PATCH] Update Version Command


diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index bd3138c00..78b41302a 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -43,8 +43,7 @@ public class VersionCommand extends BukkitCommand {
         if (!testPermission(sender)) return true;
 
         if (args.length == 0) {
-            sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")");
-            sendVersion(sender);
+            sender.sendMessage(ChatColor.translateAlternateColorCodes('&', "&cThis server is running &eFlamePaper &7(1.8.8) &cby &bLinsaFTW"));
         } else {
             StringBuilder name = new StringBuilder();
 
@@ -73,8 +72,7 @@ public class VersionCommand extends BukkitCommand {
             }
 
             if (!found) {
-                sender.sendMessage("This server is not running any plugin by that name.");
-                sender.sendMessage("Use /plugins to get a list of plugins.");
+                sender.sendMessage(ChatColor.translateAlternateColorCodes('&', "&cThis server is not running any plugin by that name.\n&cUse /plugins to get a list of plugins."));
             }
         }
         return true;
-- 
2.32.0

